NewHaven/src/TurnObserver.cpp

```diff
@@ -0,0 +1,150 @@
+//
+// Created by hyperion on 2020-04-21.
+//
+
+#include "TurnObserver.h"
+#include <iostream>
+#include "GameController.h"
+
+using namespace std;
+TurnObserver::TurnObserver():current{nullptr}{};
+TurnObserver::~TurnObserver() {
+    delete current;
+};
+void TurnObserver::printTurnStart() {
+    // At the start of a turn we should print that the player is playing a new turn
+    int player_index =  GameController::current->getCurrentPlayerTurn();
+    delete current;
+    current = (*GameController::current->game_settings->players)[player_index];
+
+    // use boost to cast our int to string
+    string msg = "\n" + current->getID() + " has started their turn!";
+    cout << msg << endl;
+}
```

Comment:
Here's the cause of all the crashing I get during run-time. More specifically lines 16-18. Let's go through this step by step assuming there are 2 players in the game:

1.
    1.1 On the first turn, we have the current player's index in `player_index`.
    1.2 We delete `current`. At that point `current` was a `nullptr` so that doesn't do anything.
    1.3 After that, we assign a pointer to `current`. So `current` starts pointing where the pointer it was assigned is pointing. That is at index `player_index` in the `players` vector of `Setting`. Now the `players` vector has pointers that point to the object of all the players of the game. Basically, `current` now point to one of those players too.

2.
    1.1 On the second time `printTurnStart()` is called, which is when the next turn is about to start, `player_index` is updated to the index of the new current player.
    2.2 Now current is deleted. Now, take a look back where I describe what `current` was pointing to. That object is completely deleted from the heap. The data of one player of the game was erased. The vector `players` from `Setting` now has a dangling pointer. The pointer at index right before the current index. So far, no issues.
    2.3 `current` now points to the object pointed to by the second pointer in the `players` vector. The turn starts normally.

3. The crash happens when the current player finishes his turn, and it's time to share the wealth. Since there are 2 players in this game, it's for the other player to build his village board. That player was deleted from the heap. Attempting anything on that player using the dangling pointer in the vector `players` will crash the game.

The thing is that at the start of each turn after the first, the player that finished his turn is removed from the game, according to the current code in `printTurnStart()`.

Reply:
shit you are right i hadnt noticed that! it is as easily resolved as removing the delete current or making a copy of it. However, considering our problems with VGMap, i would prefer avoid copying the player. I'll omitt the delete current. It is a logic mistake on my end. Really strange that i was not being affected... might be due to larger cache or memory configurations on some systems idk.